opal/mca/btl/ofi/README

```diff
@@ -0,0 +1,69 @@
+========================================
+Design notes on BTL/OFI
+========================================
+
+This is the RDMA only btl based on OFI Libfabric. The goal is to enable RDMA
+with multiple vendor hardware through one interface. Most of the operations are
+managed by upper layer (osc/rdma). This BTL is mostly doing the low level work.
+
+Tested providers: sockets,psm2,ugni
+
+========================================
+
+Component
+
+This BTL is requesting libfabric version 1.5 API and will not support older versions.
+
+The required capabilities of this BTL is FI_ATOMIC and FI_RMA with the endpoint type
+of FI_EP_RDM only. This BTL does NOT support libfabric provider that requires local
+memory registration (FI_MR_LOCAL).
+
+The provider specific settings can be added to btl_ofi_providers.h. This includes the
+additional hints that some provider might need to be funcional and also validation of
+info that returns from libfabric.
+
+The BTL will initialize a module with ONLY the first compatible info returned from OFI.
+This means it will rely on OFI provider to do load balancing. The support for multiple
+device might be added later.
+
+The BTL creates only one endpoint and one CQ.
+
+========================================
+
+Memory Registration
+
+Open MPI has a system in place to exchange remote address and always use the remote
+virtual address to refer to a piece of memory. However, some libfabric providers might
+not support the use of virtual address and instead will use zero-based offset adressing.
+
+FI_MR_VIRT_ADDR is the flag that determine this behavior. mca_btl_ofi_reg_mem() handles
+this by storing the base address in registration handle in case of the provider does not
+support FI_MR_VIRT_ADDR. This base address will be used to calculate the offset later in
+RDMA/Atomic operations.
+
+The BTL does not register local operand or compare. This is why this BTL does not support
+FI_MR_LOCAL.
+
+========================================
+
+Completions
+
+Every operation in this BTL is asynchronous. The completion handling will occur in
+mca_btl_ofi_component_progress() where we read the CQ with the completion context and
+execute the callback functions. The completions are local. No remote completion event is
+generated.
+
+The BTL keep tracks of number of outstanding operations and provide flush interface.
+
+========================================
+
+Sockets Provider
+
+Sockets provider is the proof of concept provider for libfabric. It is supposed to support
+all the OFI API with emulations. This provider is considered very slow and bound to raise
+problems that we might not see from other faster providers.
+
+Known Problems:
+    - sockets provider uses progress thread and can cause segfault in finalize as we free
+      the resources while progress thread is still using it. sleep(1) was put in
+      mca_btl_ofi_componenet_close() for this reason.
```

Comment:
Hmm.  Seems sketchy.  This sounds like a bug either in libfabric or in this BTL.

Reply:
I believe this is related to #5195 .  Will check again after we have a fix for that..